src/client/client.c

```diff
@@ -1,32 +1,180 @@
+                free(client->object_path);
+        }
+
+        free(client);
+}
+
+UnitList *new_unit_list() {
+        _cleanup_unit_list_ UnitList *unit_list = malloc0(sizeof(UnitList));
+
+        if (unit_list == NULL) {
+                return NULL;
+        }
+
+        unit_list->ref_count = 1;
+        LIST_HEAD_INIT(unit_list->units);
+
+        return steal_pointer(&unit_list);
+}
+
+UnitList *unit_list_ref(UnitList *unit_list) {
+        unit_list->ref_count++;
+        return unit_list;
+}
+
+void unit_list_unref(UnitList *unit_list) {
+        unit_list->ref_count--;
+        if (unit_list->ref_count != 0) {
+                return;
+        }
+
+        UnitInfo *unit = NULL, *next_unit = NULL;
+        LIST_FOREACH_SAFE(units, unit, next_unit, unit_list->units) {
+                unit_unref(unit);
+        }
+        free(unit_list);
+}
+
+NodesUnitList *new_nodes_unit_list() {
+        _cleanup_nodes_unit_list_ NodesUnitList *nodes_unit_list = malloc0(sizeof(NodesUnitList));
+
+        if (nodes_unit_list == NULL) {
+                return NULL;
+        }
+
+        nodes_unit_list->ref_count = 1;
+        LIST_HEAD_INIT(nodes_unit_list->nodes_units);
+
+        return steal_pointer(&nodes_unit_list);
+}
 
-static void signal_handler(int signal) {
-        printf("\n%s(): Received signal %d\n", __func__, signal);
-        done = true;
+NodesUnitList *nodes_unit_list_ref(NodesUnitList *nodes_unit_list) {
+        nodes_unit_list->ref_count++;
+        return nodes_unit_list;
 }
 
-int main() {
-        struct sigaction sa;
-        sa.sa_handler = &signal_handler;
-        sigaction(SIGINT, &sa, NULL);
-        sigaction(SIGTERM, &sa, NULL);
-        sigaction(SIGSEGV, &sa, NULL);
-        sigaction(SIGBUS, &sa, NULL);
-        sigaction(SIGABRT, &sa, NULL);
+void nodes_unit_list_unref(NodesUnitList *nodes_unit_list) {
+        nodes_unit_list->ref_count--;
+        if (nodes_unit_list->ref_count != 0) {
+                return;
+        }
 
-        while (!done) {
-                sleep(1);
-                printf("%s(): Hello from client\n", __func__);
+        UnitList *node_unit_list = NULL, *next_node_unit_list = NULL;
+        LIST_FOREACH_SAFE(node_units, node_unit_list, next_node_unit_list, nodes_unit_list->nodes_units) {
+                unit_list_unref(node_unit_list);
         }
+        free(nodes_unit_list);
+}
+
+int client_call_manager(Client *client) {
+        int r = 0;
 
-        printf("%s(): Goodbye fromt client\n", __func__);
+        r = sd_bus_open_user(&(client->user_bus));
+        if (r < 0) {
+                fprintf(stderr, "Failed to connect to system bus: %s\n", strerror(-r));
+                return r;
+        }
+
+        if (streq(client->op, "list-units")) {
+                if (client->opargc == 0) {
+                        r = list_units_on_all(client);
+                } else {
+                        r = list_units_on(client->opargv[0], client);
+                }
```

Comment:
```suggestion
                if (client->opargc == 0) {
                        return list_units_on_all(client);
                }
                return list_units_on(client->opargv[0], client);
```

Reply:
I don't know if that's better. These are two equal option, the way I see it is a bit more confusing.

Reply:
`else` encourages complexer code structure and makes it less readable. In addition, I prefer early returns. 
The option is `list-units`. These are two variants - listing them for one or all nodes - and listing them for a named node is the default. I am not sure that using `else` or not is more expressive to differentiate. 
But this is just a minor suggestion and doesn't block from merging :)